mbedtls/src/ecp/mod.rs

```diff
@@ -0,0 +1,239 @@
+            mpi_copy(&mut ret.inner.Y, y.handle()).into_result()?;
+            mpi_lset(&mut ret.inner.Z, 1).into_result()?;
+        };
+
+        Ok(ret)
+    }
+
+    pub fn x(&self) -> ::Result<Mpi> {
+        Mpi::copy(&self.inner.X)
+    }
+
+    pub fn y(&self) -> ::Result<Mpi> {
+        Mpi::copy(&self.inner.Y)
+    }
+
+    pub fn is_zero(&self) -> ::Result<bool> {
+        /*
+        mbedtls_ecp_is_zero takes arg as non-const for no particular reason
+        use this unsafe cast here to avoid having to take &mut self
+        */
+        match unsafe { ecp_is_zero(&self.inner as *const ecp_point as *mut ecp_point) } {
+            0 => Ok(false),
+            1 => Ok(true),
+            _ => Err(::Error::EcpInvalidKey),
+        }
+    }
+
+    pub fn mul(&self, group: &mut EcGroup, k: &Mpi) -> ::Result<EcPoint> {
+        // TODO provide random number generator for blinding
+        // Note: mbedtls_ecp_mul performs point validation itself so we skip that here
+
+        let mut ret = Self::init();
+
+        unsafe {
+            ecp_mul(
+                &mut group.inner,
+                &mut ret.inner,
+                k.handle(),
+                &self.inner,
+                None,
+                ::core::ptr::null_mut(),
+            )
+        }
+        .into_result()?;
+
+        Ok(ret)
+    }
+
+    /// Compute pt1*k1 + pt2*k2 -- not const time
+    pub fn muladd(
+        group: &mut EcGroup,
+        pt1: &EcPoint,
+        k1: &Mpi,
+        pt2: &EcPoint,
+        k2: &Mpi,
+    ) -> ::Result<EcPoint> {
+        let mut ret = Self::init();
+
+        if group.contains_point(&pt1)? == false {
+            return Err(::Error::EcpInvalidKey);
+        }
+
+        if group.contains_point(&pt2)? == false {
+            return Err(::Error::EcpInvalidKey);
+        }
+
+        unsafe {
+            ecp_muladd(
+                &mut group.inner,
+                &mut ret.inner,
+                k1.handle(),
+                &pt1.inner,
+                k2.handle(),
+                &pt2.inner,
+            )
+        }
+        .into_result()?;
+
+        Ok(ret)
+    }
+
+    pub fn eq(&self, other: &EcPoint) -> ::Result<bool> {
+        let r = unsafe { ecp_point_cmp(&self.inner, &other.inner) };
+
+        match r {
+            0 => Ok(true),
+            ERR_ECP_BAD_INPUT_DATA => Ok(false),
+            x => Err(::Error::from_mbedtls_code(x)),
+        }
+    }
+
+    pub fn to_binary(&self, group: &EcGroup, compressed: bool) -> ::Result<Vec<u8>> {
+        /*
+        We know biggest group supported is P-521 so just allocate a
+        vector big enough and then resize it down to the actual output
+        length
+
+        ceil(521/8) == 66
+        OS2ECP format is header byte + 2 point elements (or 1 if compressed)
+        so max size is 66*2+1 = 133
+         */
+        let mut olen = 0;
+        let mut buf = Vec::new();
+        buf.resize(133, 0u8);
```

Comment:
`let mut buf = vec![0u8; 133];`

Reply:
`vec!` is missing from the no-std builds. For some reason it is not exported by alloc crate?

Reply:
I think you need to add `#[macro_use]` on the `extern crate alloc`